### PR TITLE
Add a checkinternaldomain to the upcheck to validate internal domains

### PR DESCRIPTION
**Pull request by andrew-edgar**
This is to add validation to both upcheck domain and a bosh defined internal domain.   

I would appreciate any assistance to this PR to see if there are additional tests to be implemented and how.  As there seems to be lots that might be effected including performance and integration tests.

Let me know how you want to proceed.

### CODE REVIEW

File: jobs/bosh-dns/spec [hunk 1]
```diff
@@ -122,6 +122,10 @@ properties:
     default:
       - upcheck.bosh-dns.
 
+  internal_check_domain:
+    description: "Domain name to use to validate internal resolution is working"
+    default: "cloud-controller-ng.service.cf.internal"
```
Comment (cjnosal) [inline jobs/bosh-dns/spec:127]: I think I would prefer having each VM resolve itself (either leveraging https://bosh.io/docs/jobs/#properties-spec in the templates, or perhaps using a reverse lookup on the local IP to get the FQDN then resolving that.

bosh-dns is deployment agnostic, so a CF alias isn't an appropriate default. And if VMs are resolving another internal VM as an upcheck, that means two things:
- they'll all be polling the health server on that VM (extra load)
- when that one VM is unhealthy (e.g. being deployed) then bosh-dns will restart on every other VM, and depending on timing that could cause the deployment to fail

File: src/bosh-dns/dns/main.go [hunk 2]
```diff
@@ -182,8 +182,8 @@ func mainExitCode() int {
 	for _, upcheckDomain := range config.UpcheckDomains {
 		mux.Handle(upcheckDomain, handlers.NewUpcheckHandler(logger))
 		for _, addr := range listenAddrs {
-			upchecks = append(upchecks, server.NewDNSAnswerValidatingUpcheck(addr, upcheckDomain, "udp"))
-			upchecks = append(upchecks, server.NewDNSAnswerValidatingUpcheck(addr, upcheckDomain, "tcp"))
+			upchecks = append(upchecks, server.NewDNSAnswerValidatingUpcheck(addr, config.InternalCheckDomain, recordSet.Hosts[0].FQDN, "udp"))
```
Comment (cjnosal) [inline src/bosh-dns/dns/main.go:185]: upcheckDomain should still be checked
The current loop is checking InternalCheckDomain once per defined upcheck domain

File: src/bosh-dns/dns/server/upcheck.go [hunk 3]
```diff
@@ -65,6 +67,29 @@ func (uc DNSAnswerValidatingUpcheck) IsUp() error {
 		return uc.wrapError(errors.New("DNS upcheck does not return the correct answer"))
 	}
 
+	request = dns.Msg{
+		Question: []dns.Question{
+			{Name: uc.internalDomain, Qtype: dns.TypeA},
```
Comment (cjnosal) [inline src/bosh-dns/dns/server/upcheck.go:72]: Some environments might be pure ipv6 (TypeAAAA). A TypeAny request would be one option.

File: src/bosh-dns/dns/server/upcheck.go [hunk 4]
```diff
@@ -65,6 +67,29 @@ func (uc DNSAnswerValidatingUpcheck) IsUp() error {
...
+		},
+	}
+	msg, _, err = dnsClient.Exchange(&request, uc.target)
+
+	if err != nil {
+		return uc.wrapError(err)
+	}
+	if msg.Rcode != dns.RcodeSuccess {
+		return uc.wrapError(errors.New("DNS resolve failed"))
```
Comment (cjnosal) [inline src/bosh-dns/dns/server/upcheck.go:81]: The error messages should be disambiguated so we can tell which test failed

**Comment by beyhan** [comment]
I think that this will be an useful addition to bosh-dns and would like to help with this pull-request. I created a new pr https://github.com/cloudfoundry/bosh-dns-release/pull/59, where I tried to apply the suggestions provided by @xtreme-conor-nosal. @andrew-edgar I created the pull-request in the `bosh-deps-forks` organisation. I hope you have rights to push changes there.

File: src/bosh-dns/dns/server/upcheck.go (hunk 3 above)
Comment (aashah) [reply src/bosh-dns/dns/server/upcheck.go:72]: I don't think we have any Bosh deployments in the wild that are pure Ipv6, I don't think that would even work with bosh..

**Comment by aashah** [comment]
We merged in #59 which overlaps a lot with the intent behind your PR @andrew-edgar. Could you take a look and see if that's true?

I'm going to close this PR out for now, but let's continue the discussion if there's something we're missing!

**Comment by andrew-edgar** [comment]
yeah sounds good thanks

**Pull request closed without merging**